### ISSUE TITLE
Failing to evaluate matrix expressions 

### ISSUE THREAD

**Issue by hurak** [jump-dev/Convex.jl#447]
The function `evaluate` seems to fail when evaluating matrix expressions. Below is a minimum working example demonstrating an eigenvalue minimization for an affine matrix function of a vector argument: 

```
minimize   λ
subject to   A(x)⪯λ, where x is a vector.
```

Preparing the problem description in Convex.jl:
```julia
using Convex
using SCS

a₀ = [1.0 2.0; 2.0 3.0]
a₁ = [1.0 0.0; 0.0 -1.0]
a₂ = [4.0 5.0; 5.0 -6.0]

x = Variable(2)
λ = Variable()

A = a₀ + x[1]*a₁ + x[2]*a₂

problem = minimize(λ,A⪯λ)
```
Solving the problem:
```julia
julia> solve!(problem, SCS.Optimizer)
----------------------------------------------------------------------------
    SCS v2.1.3 - Splitting Conic Solver
    (c) Brendan O'Donoghue, Stanford University, 2012
----------------------------------------------------------------------------
Lin-sys: sparse-indirect, nnz in A = 10, CG tol ~ 1/iter^(2.00)
eps = 1.00e-05, alpha = 1.50, max_iters = 5000, normalize = 1, scale = 1.00
acceleration_lookback = 10, rho_x = 1.00e-03
Variables n = 4, constraints m = 5
Cones:  primal zero / dual free vars: 2
    sd vars: 3, sd blks: 1
Setup time: 5.91e-05s
----------------------------------------------------------------------------
 Iter | pri res | dua res | rel gap | pri obj | dua obj | kap/tau | time (s)
----------------------------------------------------------------------------
     0| 7.17e+18  8.71e+18  1.00e+00 -2.56e+19  3.48e+19  2.38e+19  3.75e-05 
    40| 2.10e-09  2.96e-09  2.08e-10  2.00e+00  2.00e+00  8.65e-18  2.06e-04 
----------------------------------------------------------------------------
Status: Solved
Timing: Solve time: 2.11e-04s
    Lin-sys: avg # CG iterations: 1.78, avg solve time: 3.58e-07s
    Cones: avg projection time: 5.61e-08s
    Acceleration: avg step time: 2.76e-06s
----------------------------------------------------------------------------
Error metrics:
dist(s, K) = 4.7016e-16, dist(y, K*) = 0.0000e+00, s'y/|s||y| = -5.4509e-01
primal res: |Ax + s - b|_2 / (1 + |b|_2) = 2.1013e-09
dual res:   |A'y + c|_2 / (1 + |c|_2) = 2.9555e-09
rel gap:    |c'x + b'y| / (1 + |c'x| + |b'y|) = 2.0849e-10
----------------------------------------------------------------------------
c'x = 2.0000, -b'y = 2.0000
============================================================================
```

Now, we can evaluate the variables:
```julia 
julia> evaluate(λ)
2.0000000053437748

julia> evaluate(x)
2-element Vector{Float64}:
 1.0000000013858472
 6.18691098963156e-10
```
We can even evaluate some general/arbitrary (unrelated to our problem statement) expression containing the variables:
```julia
julia> evaluate(1.0 + 2.0*x[1] + 3.0*x[2])
1-element Vector{Float64}:
 3.000000004627768
```

But we fail when trying to evaluate the (original) matrix expression:
```julia
julia> evaluate(A)
ERROR: DimensionMismatch("A has dimensions (1,1) but B has dimensions (2,2)")
Stacktrace:
  [1] gemm_wrapper!(C::Matrix{Float64}, tA::Char, tB::Char, A::Matrix{Float64}, B::Matrix{Float64}, _add::LinearAlgebra.MulAddMul{true, true, Bool, Bool})
    @ LinearAlgebra /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:643
  [2] mul!
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:169 [inlined]
  [3] mul!
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:275 [inlined]
  [4] *
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:160 [inlined]
  [5] *(a::Vector{Float64}, B::Matrix{Float64})
    @ LinearAlgebra /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:63
  [6] evaluate(x::Convex.MultiplyAtom)
    @ Convex ~/.julia/packages/Convex/Xothw/src/atoms/affine/multiply_divide.jl:53
  [7] _mapreduce(f::typeof(evaluate), op::Convex.var"#16#17", #unused#::IndexLinear, A::Vector{Convex.AbstractExpr})
    @ Base ./reduce.jl:408
  [8] _mapreduce_dim
    @ ./reducedim.jl:318 [inlined]
  [9] #mapreduce#672
    @ ./reducedim.jl:310 [inlined]
 [10] mapreduce
    @ ./reducedim.jl:310 [inlined]
 [11] evaluate(x::Convex.AdditionAtom)
    @ Convex ~/.julia/packages/Convex/Xothw/src/atoms/affine/add_subtract.jl:99
 [12] top-level scope
    @ none:1
```
 
It is not perfectly clear from the documentation (the [section on Expressions](https://jump.dev/Convex.jl/stable/types/#Expressions)) if evaluation of matrix expressions is supported. But since these expressions are valid for the problem description, their evaluation should be supported.

**Comment by ericphanson** [comment]
This definitely should work; from the stacktrace I’d guess there’s a bug in

```julia
  [6] evaluate(x::Convex.MultiplyAtom)
    @ Convex ~/.julia/packages/Convex/Xothw/src/atoms/affine/multiply_divide.jl:53
  [7] _mapreduce(f::typeof(evaluate), op::Convex.var"#16#17", #unused#::IndexLinear, A::Vector{Convex.AbstractExpr})
```
although I haven’t taken a look yet. If you’re up for digging in, any help would be great!

**Comment by hurak** [comment]
I will try to help find and possibly even correct the bug. But we will see if I stand any chance.

First, I at least isolated the problem a bit more. At appears that it is not the "matrixness" of the expression `A(x) = a₀ + a₁x₁ + a₂x₂` that causes the problem but rather the "vectorness" of the variable `x`. If I modified the optimization problem such that `x` is just a scalar as in

```julia
using Convex
using SCS

a₀ = [1.0 2.0; 2.0 3.0]
a₁ = [1.0 0.0; 0.0 -1.0]

x = Variable()
λ = Variable()

A = a₀ + x*a₁

problem = minimize(λ,A⪯λ)
```
```julia
julia> solve!(problem,SCS.Optimizer)
----------------------------------------------------------------------------
    SCS v2.1.3 - Splitting Conic Solver
    (c) Brendan O'Donoghue, Stanford University, 2012
----------------------------------------------------------------------------
Lin-sys: sparse-indirect, nnz in A = 7, CG tol ~ 1/iter^(2.00)
eps = 1.00e-05, alpha = 1.50, max_iters = 5000, normalize = 1, scale = 1.00
acceleration_lookback = 10, rho_x = 1.00e-03
Variables n = 3, constraints m = 5
Cones:  primal zero / dual free vars: 2
    sd vars: 3, sd blks: 1
Setup time: 6.23e-05s
----------------------------------------------------------------------------
 Iter | pri res | dua res | rel gap | pri obj | dua obj | kap/tau | time (s)
----------------------------------------------------------------------------
     0| 1.50e+19  1.53e+19  1.00e+00 -3.42e+19  6.36e+19  1.38e+19  2.36e-05 
    20| 6.19e-11  7.33e-11  5.58e-12  2.00e+00  2.00e+00  1.83e-16  1.42e-04 
----------------------------------------------------------------------------
Status: Solved
Timing: Solve time: 1.49e-04s
    Lin-sys: avg # CG iterations: 1.33, avg solve time: 4.30e-07s
    Cones: avg projection time: 8.79e-08s
    Acceleration: avg step time: 3.85e-06s
----------------------------------------------------------------------------
Error metrics:
dist(s, K) = 1.4487e-16, dist(y, K*) = 0.0000e+00, s'y/|s||y| = -4.2600e-01
primal res: |Ax + s - b|_2 / (1 + |b|_2) = 6.1870e-11
dual res:   |A'y + c|_2 / (1 + |c|_2) = 7.3335e-11
rel gap:    |c'x + b'y| / (1 + |c'x| + |b'y|) = 5.5833e-12
----------------------------------------------------------------------------
c'x = 2.0000, -b'y = 2.0000
============================================================================
```
the problem disappears:
```julia
julia> evaluate(λ)
1.9999999999215436

julia> evaluate(x)
1.0000000000139493

julia> evaluate(A)
2×2 Matrix{Float64}:
 2.0  2.0
 2.0  2.0
```

While debugging the original problem, it appears that the culprits are the atoms corresponding to the two summands `a₁x₁` and `a₂x₂` (a reminder: `a₁` and `a₂` are matrices, `x₁` and `x₂` are scalar components of a vector `x`). 

```julia
julia> Juno.@enter evaluate(A)
debug> x
+ (affine; real)
├─ 2×2 Matrix{Float64}
├─ * (affine; real)
│  ├─ index (affine; real)
│  │  └─ 2-element real variable (id: 774…508)
│  └─ 2×2 Matrix{Float64}
└─ * (affine; real)
   ├─ index (affine; real)
   │  └─ 2-element real variable (id: 774…508)
   └─ 2×2 Matrix{Float64}

debug> x.children[2]
* (affine; real)
├─ index (affine; real)
│  └─ 2-element real variable (id: 774…508)
└─ 2×2 Matrix{Float64}
```
In particular, it appears that the atom corresponding to the second summand, namely `a₁x₁`, carries inside the information that the real variable has two components - somehow it keeps inside the information about the whole vector `x` even it it only depends on `x₁`. Cannot this be the reason for the troubles? If I now call `evaluate` on this atom, the code errors:

```julia
debug> evaluate(x.children[2])
ERROR: DimensionMismatch("A has dimensions (1,1) but B has dimensions (2,2)")
Stacktrace:
  [1] gemm_wrapper!(C::Matrix{Float64}, tA::Char, tB::Char, A::Matrix{Float64}, B::Matrix{Float64}, _add::LinearAlgebra.MulAddMul{true, true, Bool, Bool})
    @ LinearAlgebra /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:643
  [2] mul!
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:169 [inlined]
  [3] mul!
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:275 [inlined]
  [4] *
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:160 [inlined]
  [5] *(a::Vector{Float64}, B::Matrix{Float64})
    @ LinearAlgebra /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:63
  [6] evaluate(x::Convex.MultiplyAtom)
    @ Convex ~/.julia/dev/Convex/src/atoms/affine/multiply_divide.jl:53
  [7] top-level scope
    @ none:1
  [8] eval
    @ ./boot.jl:360 [inlined]
  [9] eval_code(frame::JuliaInterpreter.Frame, expr::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/kFbUU/src/utils.jl:649
 [10] eval_code(frame::JuliaInterpreter.Frame, command::String)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/kFbUU/src/utils.jl:627
 [11] interpret
    @ ~/.julia/packages/Atom/5NgzM/src/debugger/repl.jl:96 [inlined]
 [12] interpret
    @ ~/.julia/packages/Atom/5NgzM/src/debugger/repl.jl:95 [inlined]
 [13] (::Atom.JunoDebugger.var"#67#68")(s::REPL.LineEdit.MIState, buf::IOBuffer, ok::Bool)
    @ Atom.JunoDebugger ~/.julia/packages/Atom/5NgzM/src/debugger/repl.jl:39
 [14] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
 [15] invokelatest
    @ ./essentials.jl:706 [inlined]
 [16] run_interface(terminal::REPL.Terminals.TextTerminal, m::REPL.LineEdit.ModalInterface, s::REPL.LineEdit.MIState)
    @ REPL.LineEdit /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/LineEdit.jl:2441
 [17] run_interface(terminal::REPL.Terminals.TextTerminal, m::REPL.LineEdit.ModalInterface)
    @ REPL.LineEdit /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/LineEdit.jl:2435
 [18] debugprompt()
    @ Atom.JunoDebugger ~/.julia/packages/Atom/5NgzM/src/debugger/repl.jl:55
 [19] (::Atom.JunoDebugger.var"#55#57")()
    @ Atom.JunoDebugger ./task.jl:411
```
No problems to `evaluate` the children's children, which in this case are `x[1]` and `a₁`:

```julia
debug> evaluate(x.children[2].children[1])
1-element Vector{Float64}:
 1.0000000013858472

debug> evaluate(x.children[2].children[2])
2×2 Matrix{Float64}:
 1.0   0.0
 0.0  -1.0
```
It really appears that it is the evaluation of the product of these two (a scalar component of a vector and a matrix) that leads to an error.

As soon as `x` is just a scalar (as I did in the simplified example above), the troubles are gone.

That is what I have so far.

**Comment by ericphanson** [comment]
Ahh I think that's exactly the issue. Convex.jl has made the choice to treat everything as a matrix (I suspect this choice as a carryover from MATLAB but I don't know the origin of it), so scalars are internally treated as 1x1 matrices. We have the function `output` to convert these back to scalars when returning things to the user, but I think that isn't being used in `IndexAtom` when it should be. I believe then the fix is to call `output` on the result of `getindex` here:

https://github.com/jump-dev/Convex.jl/blob/52e4eab860a5b006462fe23295e14d282c994740/src/atoms/affine/index.jl#L41-L47

then will cause the 1x1 matrices to be correctly represented as scalars so the multiplication succeeds. Would you be up for making a PR?

The steps would be to

* add an `output` call to the two branches of that function,
* add a comment there pointing to this issue,
* and add a test with your original example

The tests are written in a particular way so that they can be re-used as correctness tests for the solvers themselves, so they are actually written as functions in our ProblemDepot module; the [docs](https://jump.dev/Convex.jl/dev/problem_depot/#How-to-write-a-ProblemDepot-problem) discuss how to add one.

**Comment by hurak** [comment]
I will certainly try as it seems doable for me with the instructions you gave. I will either make it right today, and if not, then in a week time or so.

**Comment by ericphanson** [comment]
Closed by #448
